src/Conversion/ONNXToTOSA/Math/Gemm.cpp

```diff
@@ -0,0 +1,208 @@
+    int64_t transB = adaptor.transB();
+    FloatAttr alpha = adaptor.alphaAttr();
+    FloatAttr beta = adaptor.betaAttr();
+    auto AType = A.getType().cast<RankedTensorType>();
+    auto BType = B.getType().cast<RankedTensorType>();
+    auto shapeA = A.getType().dyn_cast<ShapedType>().getShape();
+    auto shapeB = B.getType().dyn_cast<ShapedType>().getShape();
+    auto resultType = getTypeConverter()->convertType(op.getResult().getType()).cast<TensorType>();
+    
+    // C is optional, if it's not there, we need to be aware of it for later computations
+    bool isCPresent = C.getType().isa<TensorType>();
+    // If it's not present, ONNX creates a NoValue operator. Remove it.
+    if (!isCPresent) {
+      rewriter.eraseOp(C.getDefiningOp());
+    }
+
+    // If legal, create a FullyConnected operator instead
+    if (checkLegalTosaFCOp(op, adaptor)) {
+      return rewriteToTosaFC(op, adaptor, rewriter);
+    }
+    
+    // ONNX gives 2d matrix as input and expect a 2d output. TOSA expects everything to be 3D. As such, there
+    // is a need to add reshapes operators before and after we do any computation. We set the batch as 1 as it
+    // is unknown.
+    llvm::SmallVector<int64_t> newShapeA{1};
+    llvm::SmallVector<int64_t> newShapeB{1};
+    newShapeA.append({shapeA[0], shapeA[1]});
+    newShapeB.append({shapeB[0], shapeB[1]});
+
+    A = tosa::CreateOpAndInfer<tosa::ReshapeOp>(
+        rewriter, op->getLoc(),
+        UnrankedTensorType::get(AType.getElementType()), A,
+        rewriter.getI64ArrayAttr(newShapeA)).getResult();
+    B = tosa::CreateOpAndInfer<tosa::ReshapeOp>(
+        rewriter, op->getLoc(),
+        UnrankedTensorType::get(BType.getElementType()), B,
+        rewriter.getI64ArrayAttr(newShapeB)).getResult();
+
+    auto tosaResult = UnrankedTensorType::get(resultType.getElementType());
+
+    // If transA or transB are present, create Transpose operators.
+    if (transA) {
+      Value targetTensor = mlir::tosa::getConstTensor<int32_t>(rewriter, op, {0, 2, 1}, {3}).value();
+      Type outputType = UnrankedTensorType::get(AType.getElementType());
+      A = tosa::CreateOpAndInfer<tosa::TransposeOp>(rewriter, loc, outputType, A, targetTensor).getResult();
+    }
+    if (transB) {
+      Value targetTensor = mlir::tosa::getConstTensor<int32_t>(rewriter, op, {0, 2, 1}, {3}).value();
+      Type outputType = UnrankedTensorType::get(AType.getElementType());
+      B = tosa::CreateOpAndInfer<tosa::TransposeOp>(rewriter, loc, outputType, B, targetTensor).getResult();
+    }
+    
+    Value alphaMulResult = NULL;
+    Value betaMulResult = NULL;
+    // If Alpha is present and not 1, we create a multiply operation for alpha * A
+    if (alpha && alpha.getValueAsDouble() != 1.) {
+      alphaMulResult = tosa::CreateOpAndInfer<tosa::MulOp>(rewriter, loc,
+            tosaResult,
+            tosa::getTosaConstTensorSingleF32(rewriter, op, alpha.getValueAsDouble(), newShapeA),
```

Comment:
@maxbartel - was it in places like this that you had trouble with `double` and `F32`?

Reply:
Yes exactly. This line works because we moved the double to float cast to the function arguments.